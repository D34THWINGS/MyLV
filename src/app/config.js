module.exports = process.env.CONFIG
