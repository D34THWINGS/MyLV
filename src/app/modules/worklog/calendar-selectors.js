import moment from 'moment'
import { createSelector } from 'reselect'

export const calendarYearSelector = state => state.calendar.year
export const calendarMonthSelector = state => state.calendar.month

export const worklogEntriesSelector = state => state.worklog.entries

export const calendarDaysSelector = createSelector(
  calendarYearSelector,
  calendarMonthSelector,
  (year, month) => {
    const m = moment(`${year}-${month}`, 'YYYY-MM')
    const weeks = []

    let w
    for (let i = 1; i <= m.daysInMonth(); i++) {
      const d = m.clone().date(i)
      if (!w || d.day() === 1) {
        w = Array.from({ length: 7 }, () => undefined)
        weeks.push(w)
      }
      w[(d.day() + 6) % 7] = `0${i}`.slice(-2)
    }

    return weeks
  },
)

export const calendarLabelsSelector = createSelector(
  worklogEntriesSelector,
  calendarYearSelector,
  calendarMonthSelector,
  (entries = {}, year, month) => {
    const dateRegExp = new RegExp(`^${year}-${month}`)
    const labelsSet = Object
      .keys(entries)
      .reduce((acc, date) => (entries[date] && dateRegExp.test(date) ? acc.add(entries[date]) : acc), new Set())
    return Array.from(labelsSet)
  },
)

export const calendarExpectedDaysSelector = createSelector(
  worklogEntriesSelector,
  calendarYearSelector,
  calendarMonthSelector,
  (entries = {}, year, month) => {
    const dateRegExp = new RegExp(`^${year}-${month}`)
    return Object.keys(entries).filter(i => entries[i] && dateRegExp.test(i)).length / 2
  },
)
