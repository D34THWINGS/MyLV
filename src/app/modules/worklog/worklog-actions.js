import { fetchWithAuth } from '../auth/auth.actions'

export const WORKLOG_FILL_MORNING = 'WORKLOG_FILL_MORNING'
export const WORKLOG_FILL_AFTERNOON = 'WORKLOG_FILL_AFTERNOON'
export const WORKLOG_FILL_DAY = 'WORKLOG_FILL_DAY'
export const WORKLOG_FILL_WEEK = 'WORKLOG_FILL_WEEK'
export const WORKLOG_FILL_MONTH = 'WORKLOG_FILL_MONTH'
export const WORKLOG_FILL_RANGE = 'WORKLOG_FILL_RANGE'
export const WORKLOG_EMPTY_DAY = 'WORKLOG_EMPTY_DAY'
export const WORKLOG_ADD_LABEL = 'WORKLOG_ADD_LABEL'
export const WORKLOG_SAVE_SUCCESS = 'WORKLOG_SAVE_SUCCESS'
export const WORKLOG_GET_START = 'WORKLOG_GET_START'
export const WORKLOG_GET_SUCCESS = 'WORKLOG_GET_SUCCESS'
export const WORKLOG_GET_ERROR = 'WORKLOG_GET_ERROR'

export const saveWorklog = (preSaveAction) => (dispatch, getState) => {
  if (preSaveAction) {
    dispatch(preSaveAction)
  }

  const { worklog: { pending } } = getState()

  if (Object.keys(pending).length === 0) {
    return Promise.resolve()
  }

  return dispatch(fetchWithAuth('/api/worklog', {
    method: 'PUT',
    body: Object.entries(pending).map(([date, label]) => ({ date, label })),
  }))
    .then(() => dispatch({ type: WORKLOG_SAVE_SUCCESS }))
}

export const fillMorning = (date, label) => saveWorklog({
  type: WORKLOG_FILL_MORNING,
  payload: {date, label},
})

export const fillAfternoon = (date, label) => saveWorklog({
  type: WORKLOG_FILL_AFTERNOON,
  payload: {date, label},
})

export const fillDay = (day, label) => saveWorklog({
  type: WORKLOG_FILL_DAY,
  payload: {day, label},
})

export const fillWeek = (day, label) => saveWorklog({
  type: WORKLOG_FILL_WEEK,
  payload: {day, label},
})

export const fillMonth = (month, label) => saveWorklog({
  type: WORKLOG_FILL_MONTH,
  payload: {month, label},
})

export const fillRange = (start, end, label) => saveWorklog({
  type: WORKLOG_FILL_RANGE,
  payload: {start, end, label},
})

export const emptyDay = (day) => saveWorklog({
  type: WORKLOG_EMPTY_DAY,
  payload: {day},
})

export const addLabel = (label, color) => ({
  type: WORKLOG_ADD_LABEL,
  payload: {label, color},
})

export const getWorklog = (year, month) => (dispatch) => {
  dispatch({ type: WORKLOG_GET_START })

  return dispatch(fetchWithAuth(`/api/worklog?year=${year}&month=${month}`))
    .then(entries => dispatch({ type: WORKLOG_GET_SUCCESS, payload: entries }))
    .catch(() => dispatch({ type: WORKLOG_GET_ERROR }))
}
