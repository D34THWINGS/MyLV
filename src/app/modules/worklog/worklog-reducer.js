import moment from 'moment'

import { publicHolidays } from './calendar-constants'
import {
  WORKLOG_FILL_MORNING,
  WORKLOG_FILL_AFTERNOON,
  WORKLOG_FILL_DAY,
  WORKLOG_FILL_WEEK,
  WORKLOG_FILL_MONTH,
  WORKLOG_EMPTY_DAY,
  WORKLOG_ADD_LABEL,
  WORKLOG_SAVE_SUCCESS
} from './worklog-actions'

const initialState = {
  entries: {},
  pending: {},
  labels: {
    'Production': '#ff9999',
    'Contribution': '#99ff99',
    'Conges payes': '#9999ff',
    'Conges sans solde': '#ff99ff',
    'Maladie': '#ffff99',
    'Autre': '#99ffff'
  }
}

const setEntry = (state, date, label) => ({
  ...state,
  entries: {
    ...state.entries,
    [date]: label
  },
  pending: {
    ...state.pending,
    [date]: label
  }
})

const setDay = (state, day, label) =>
  setEntry(setEntry(state, `${day}-am`, label), `${day}-pm`, label)

export default function (state = initialState, { type, payload }) {
  switch (type) {
    case WORKLOG_FILL_MORNING:
      return setEntry(state, `${payload.date}-am`, payload.label)
    case WORKLOG_FILL_AFTERNOON:
      return setEntry(state, `${payload.date}-pm`, payload.label)
    case WORKLOG_FILL_DAY:
      return setDay(state, payload.day, payload.label)
    case WORKLOG_FILL_WEEK:
      const startDate = moment(payload.day).startOf('week').subtract(1, 'day')
      return new Array(5)
        .fill(0)
        .reduce((s) => {
          startDate.add(1, 'day')
          if (publicHolidays.has(startDate.format('MM-DD'))) {
            return s
          }
          return setDay(s, startDate.format('YYYY-MM-DD'), payload.label)
        }, state)
    case WORKLOG_FILL_MONTH:
      return new Array(moment(`${payload.month}-01`).daysInMonth())
        .fill(0)
        .map((v, i) => i + 1)
        .filter((v) => {
          const stringDate = `${payload.month}-${`0${v}`.slice(-2)}`
          return [0, 6].indexOf(moment(stringDate).day()) === -1 && !publicHolidays.has(stringDate.slice(5))
        })
        .reduce((s, v) => setDay(s, `${payload.month}-${`0${v}`.slice(-2)}`, payload.label), state)
    case WORKLOG_EMPTY_DAY:
      return setDay(state, payload.day)
    case WORKLOG_ADD_LABEL:
      return {
        ...state,
        labels: {
          ...state.labels,
          [payload.label]: payload.color
        }
      }
    case WORKLOG_SAVE_SUCCESS:
      return {
        ...state,
        pending: {}
      }
    default:
      return state
  }
}
