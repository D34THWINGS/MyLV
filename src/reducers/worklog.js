import {
  WORKLOG_FILL_MORNING,
  WORKLOG_FILL_AFTERNOON,
  WORKLOG_FILL_DAY,
  WORKLOG_FILL_WEEK,
  WORKLOG_FILL_MONTH
} from '../actions/worklog-actions'

const initialState = {
  entries: {}
}

const setEntry = (state, date, label) => ({
  ...state,
  entries: {
    ...state.entries,
    [date]: label
  }
})

export default function (state = initialState, { type, payload }) {
  switch (type) {
    case WORKLOG_FILL_MORNING:
      return setEntry(state, payload.date, payload.label)
    case WORKLOG_FILL_AFTERNOON:
      return setEntry(state, payload.date, payload.label)
    case WORKLOG_FILL_DAY:
      return state
    case WORKLOG_FILL_WEEK:
      return state
    case WORKLOG_FILL_MONTH:
      return state
    default:
      return state
  }
}
