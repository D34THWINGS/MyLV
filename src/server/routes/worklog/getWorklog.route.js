const Joi = require('joi')
const Boom = require('boom')

module.exports = {
  method: 'GET',
  path: '/api/worklog',
  config: {
    auth: false,
    validate: {
      query: {
        year: Joi.string().required(),
        month: Joi.string().required(),
      },
    },
  },
  handler (req, res) {
    const { Entry } = req.server.app.models

    return Entry.find({
      date: { $regex: `${req.query.year}-${req.query.month}-` },
    })
      .then(entries => res.mongodb(entries, ['client', 'manager', 'id', 'userId']))
      .catch(err => res(Boom.wrap(err)))
  },
}
