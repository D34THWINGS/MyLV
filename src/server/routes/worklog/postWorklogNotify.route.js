const config = require('config')
const moment = require('moment')

const hasRole = require('../../helpers/hasRole.pre')
const getWorkingDays = require('../../helpers/getWorkingDays.helper')
const lvConnect = require('../../helpers/lvconnect.helper')

module.exports = {
  method: 'POST',
  path: '/api/worklog/notify',
  config: {
    pre: [hasRole(config.cracra.partnersRoles)],
  },
  async handler(req, res) {
    const date = moment().format('YYYY-MM')
    const entries = await req.server.app.models.Entry.find({
      date: { $regex: date },
    })

    const expectedDays = getWorkingDays(date)
    const partners = new Map()
    entries.forEach(entry => partners.set(entry.userId.toString(), (partners.get(entry.userId.toString()) || 0) + 1))
    const filled = Array.from(partners.entries())
      .filter(([, count]) => count / 2 >= expectedDays)
      .map(([userId]) => userId)

    const badBoys = await req.server.app.models.Profile.find({ userId: { $nin: filled } })
    const badBoysQuery = badBoys.map(profile => `ids=${profile.userId.toString()}`)
    const { results: reelBadBoys } = await lvConnect.api(`/users?${badBoysQuery.join('&')}`).then(res => res.json())

    await req.server.methods.sendPushNotification(reelBadBoys.map(partner => partner.id), JSON.stringify({
      message: 'CRA CRA CRA CRA !',
    }))

    res({ notified: badBoys.length })
  },
}
