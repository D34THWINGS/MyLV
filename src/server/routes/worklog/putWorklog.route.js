const Joi = require('joi')
const Boom = require('boom')
const moment = require('moment')

module.exports = {
  method: 'PUT',
  path: '/api/worklog',
  config: {
    validate: {
      payload: Joi.array().items(Joi.object().keys({
        date: Joi.string().regex(/^\d{4}-(0?[1-9]|1[012])-(0?[1-9]|[12][0-9]|3[01])-(am|pm)$/).required(),
        client: Joi.string(),
        manager: Joi.string(),
        label: Joi.alternatives().try(Joi.any().allow(null), Joi.string()),
      })),
    },
  },
  handler (req, res) {
    let isOutOfBounds
    let missingClientOrManager
    req.payload.forEach(({ date, client, manager }) => {
      const diff = moment.duration(moment(date.slice(0, -3)).diff(moment())).asMonths()
      if (Math.abs(diff) > 2) {
        isOutOfBounds = true
      }

      if ((client && !manager) || (!client && manager)) {
        missingClientOrManager = true
      }
    })

    if (isOutOfBounds) {
      return res(Boom.badRequest('One or more entries have dates farther than to months from now'))
    }

    if (missingClientOrManager) {
      return res(Boom.badRequest('One or more entries have specified a client without a manager'))
    }

    const entriesToAdd = req.payload
      .map(entry => Object.assign(entry, { userId: req.auth.credentials.id }))
      .filter(entry => !!entry.label)

    return req.server.app.models.Entry.deleteMany({
      userId: req.auth.credentials.id,
      date: { $in: req.payload.map(entry => entry.date) },
    })
      .then(() => {
        if (entriesToAdd.length > 0) {
          return req.server.app.models.Entry.create(entriesToAdd)
        }
      })
      .then(() => res({ success: true }))
      .catch(err => res(Boom.wrap(err)))
  },
}
