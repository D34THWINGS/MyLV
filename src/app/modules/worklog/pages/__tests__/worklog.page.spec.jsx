import React from 'react'
import { shallow } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'

import { WorklogPage } from '../worklog.page'
import config from '../../../../config'

jest.mock('../../../../config', () => ({}))
jest.unmock('../worklog.page')

describe('WorklogPage', () => {
  let props

  const getWrapper = () => shallow(<WorklogPage {...props} />)

  beforeEach(() => {
    props = {
      classes: {},
      shouldRemindProcess: false,
      canPrint: false,
      push: jest.fn(),
      disableProofOfTransportDialog: jest.fn(),
      shouldDisplayProofOfTransportDialog: false,
      hasInvalidTransportProof: false,
    }

    config.featureFlipping = {
      transport: false,
    }
  })

  it('should render properly', () => {
    // When
    const wrapper = getWrapper()

    // Then
    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })

  it('should render transport dialog', () => {
    // Given
    props.hasInvalidTransportProof = true
    props.shouldDisplayProofOfTransportDialog = true
    config.featureFlipping.transport = true

    // When
    const wrapper = getWrapper()

    // Then
    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })

  it('should display process reminder', () => {
    // Given
    props.shouldRemindProcess = true

    // When
    const wrapper = getWrapper()

    // Then
    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })
})
