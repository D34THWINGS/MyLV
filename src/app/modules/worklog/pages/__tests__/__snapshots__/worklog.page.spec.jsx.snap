// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorklogPage should display process reminder 1`] = `
<React.Fragment>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
  >
    <title>
      Mon CRA | 
    </title>
  </HelmetWrapper>
  <WithStyles(GridWrapper)
    container={true}
  >
    <Connect(Process) />
    <Connect(WithStyles(ReduxForm)) />
    <Connect(WithStyles(Calendar)) />
    <WithStyles(Button)
      className=""
      color="primary"
      onClick={[Function]}
      variant="fab"
    >
      <pure(Print) />
    </WithStyles(Button)>
  </WithStyles(GridWrapper)>
  <Connect(WithStyles(Printer)) />
</React.Fragment>
`;

exports[`WorklogPage should redirect if missing client info 1`] = `
<Redirect
  push={false}
  to="/client"
/>
`;

exports[`WorklogPage should render properly 1`] = `
<React.Fragment>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
  >
    <title>
      Mon CRA | 
    </title>
  </HelmetWrapper>
  <WithStyles(GridWrapper)
    container={true}
  >
    <Connect(WithStyles(ReduxForm)) />
    <Connect(WithStyles(Calendar)) />
    <WithStyles(Button)
      className=""
      color="primary"
      onClick={[Function]}
      variant="fab"
    >
      <pure(Print) />
    </WithStyles(Button)>
  </WithStyles(GridWrapper)>
  <Connect(WithStyles(Printer)) />
</React.Fragment>
`;
